File: extensions/tiktok/src/main/java/app/morphe/extension/tiktok/spoof/sim/SpoofSimPatch.java
/*
 * Forked from:
 * https://gitlab.com/ReVanced/revanced-patches/-/blob/main/extensions/tiktok/src/main/java/app/revanced/extension/tiktok/spoof/sim/SpoofSimPatch.java
 */

package app.morphe.extension.tiktok.spoof.sim;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.Utils;
import app.morphe.extension.tiktok.settings.Settings;
import app.morphe.extension.tiktok.settings.SettingsStatus;

@SuppressWarnings("unused")
public class SpoofSimPatch {
    // Values are replaced by the patch-time "Bypass regional restrictions" option.
    public static boolean DEFAULT_REGION_SPOOF_ENABLED = false;
    public static String DEFAULT_REGION_ISO = "ae";
    public static String DEFAULT_REGION_MCC_MNC = "42403";
    public static String DEFAULT_REGION_OPERATOR = "du";

    private static boolean isContextNotSet(String fieldSpoofed) {
        if (Utils.getContext() != null) {
            return false;
        }

        Logger.printException(() -> "Context is not yet set, cannot spoof: " + fieldSpoofed, null);
        return true;
    }

    public static String getInitialCountryIso() {
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_ISO : "us";
    }

    public static String getInitialMccMnc() {
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_MCC_MNC : "310260";
    }

    public static String getInitialOperatorName() {
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_OPERATOR : "T-Mobile";
    }

    private static boolean useInAppRegion() {
        return SettingsStatus.simSpoofEnabled && Settings.SIM_SPOOF.get();
    }

    private static String selectedCountryIso() {
        if (useInAppRegion()) return Settings.SIM_SPOOF_ISO.get();
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_ISO : null;
    }

    private static String selectedMccMnc() {
        if (useInAppRegion()) return Settings.SIMSPOOF_MCCMNC.get();
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_MCC_MNC : null;
    }

    private static String selectedOperatorName() {
        if (useInAppRegion()) return Settings.SIMSPOOF_OP_NAME.get();
        return DEFAULT_REGION_SPOOF_ENABLED ? DEFAULT_REGION_OPERATOR : null;
    }

    public static String getCountryIso(String value) {
        if (isContextNotSet("countryIso")) return value;

        String iso = selectedCountryIso();
        if (iso == null || !iso.matches("(?i)[a-z]{2}")) return value;

        Logger.printDebug(() -> "Spoofing countryIso from: " + value + " to: " + iso);
        return iso;
    }

    public static String getOperator(String value) {
        if (isContextNotSet("MCC-MNC")) return value;

        String mccMnc = selectedMccMnc();
        if (mccMnc == null || !mccMnc.matches("[0-9]{5,6}")) return value;

        Logger.printDebug(() -> "Spoofing sim MCC-MNC from: " + value + " to: " + mccMnc);
        return mccMnc;
    }

    public static String getOperatorName(String value) {
        if (isContextNotSet("operatorName")) return value;

        String operator = selectedOperatorName();
        if (operator == null || operator.trim().isEmpty()) return value;

        Logger.printDebug(() -> "Spoofing sim operatorName from: " + value + " to: " + operator);
        return operator;
    }

    public static String getMcc(String value) {
        return getMccMncPart(value, true);
    }

    public static String getMnc(String value) {
        return getMccMncPart(value, false);
    }

    private static String getMccMncPart(String value, boolean mcc) {
        if (isContextNotSet(mcc ? "cellMcc" : "cellMnc")) return value;
        String combined = selectedMccMnc();
        if (combined == null || !combined.matches("[0-9]{5,6}")) return value;

        String replacement = mcc ? combined.substring(0, 3) : combined.substring(3);
        Logger.printDebug(() -> "Spoofing " + (mcc ? "cell MCC" : "cell MNC")
                + " from: " + value + " to: " + replacement);
        return replacement;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/tiktok/misc/spoof/sim/SpoofSimPatch.kt
/*
 * Forked from:
 * https://gitlab.com/ReVanced/revanced-patches/-/blob/main/patches/src/main/kotlin/app/revanced/patches/tiktok/misc/spoof/sim/SpoofSimPatch.kt
 */

package app.morphe.patches.tiktok.misc.spoof.sim

import app.morphe.patcher.extensions.InstructionExtensions.addInstruction
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patcher.patch.stringOption
import app.morphe.patcher.util.proxy.mutableTypes.encodedValue.MutableBooleanEncodedValue.Companion.toMutable
import app.morphe.patcher.util.proxy.mutableTypes.encodedValue.MutableStringEncodedValue
import app.morphe.patches.shared.compat.AppCompatibilities
import app.morphe.patches.tiktok.misc.extension.sharedExtensionPatch
import app.morphe.patches.tiktok.misc.settings.SettingsStatusLoadFingerprint
import app.morphe.patches.tiktok.misc.settings.settingsPatch
import app.morphe.util.findMutableMethodOf
import app.morphe.util.getReference
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.Method
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.reference.MethodReference
import com.android.tools.smali.dexlib2.immutable.value.ImmutableBooleanEncodedValue
import com.android.tools.smali.dexlib2.immutable.value.ImmutableStringEncodedValue

private const val EXTENSION_CLASS_DESCRIPTOR = "Lapp/morphe/extension/tiktok/spoof/sim/SpoofSimPatch;"

private val regionSpoofHooksPatch = bytecodePatch(
    description = "Injects the shared TikTok region identity hooks.",
) {
    dependsOn(
        sharedExtensionPatch,
    )

    compatibleWith(*AppCompatibilities.tiktok4623())

    execute {
        val replacements = mapOf(
            "getSimCountryIso" to "getCountryIso",
            "getNetworkCountryIso" to "getCountryIso",
            "getSimOperator" to "getOperator",
            "getNetworkOperator" to "getOperator",
            "getSimOperatorName" to "getOperatorName",
            "getNetworkOperatorName" to "getOperatorName",
        )

        val patchesByMethod = linkedMapOf<Method, ArrayDeque<Pair<Int, String>>>()
        classDefForEach { classDef ->
            for (method in classDef.methods) {
                val implementation = method.implementation ?: continue
                implementation.instructions.forEachIndexed { index, instruction ->
                    if (
                        instruction.opcode != Opcode.INVOKE_VIRTUAL &&
                        instruction.opcode != Opcode.INVOKE_VIRTUAL_RANGE
                    ) return@forEachIndexed

                    val methodReference = instruction.getReference<MethodReference>()
                        ?: return@forEachIndexed
                    val replacement = when {
                        methodReference.definingClass == "Landroid/telephony/TelephonyManager;" ->
                            replacements[methodReference.name]
                        methodReference.definingClass.startsWith("Landroid/telephony/CellIdentity") &&
                            methodReference.name == "getMccString" -> "getMcc"
                        methodReference.definingClass.startsWith("Landroid/telephony/CellIdentity") &&
                            methodReference.name == "getMncString" -> "getMnc"
                        else -> null
                    }
                    if (replacement == null || methodReference.returnType != "Ljava/lang/String;") {
                        return@forEachIndexed
                    }
                    patchesByMethod.getOrPut(method) { ArrayDeque() }.add(index to replacement)
                }
            }
        }

        patchesByMethod.forEach { (method, patches) ->
            val mutableMethod = mutableClassDefBy(method.definingClass).findMutableMethodOf(method)
            while (patches.isNotEmpty()) {
                val (index, replacement) = patches.removeLast()
                val resultRegister = mutableMethod.getInstruction<OneRegisterInstruction>(index + 1).registerA

                mutableMethod.addInstructions(
                    index + 2,
                    """
                        invoke-static { v$resultRegister }, $EXTENSION_CLASS_DESCRIPTOR->$replacement(Ljava/lang/String;)Ljava/lang/String;
                        move-result-object v$resultRegister
                    """,
                )
            }
        }

    }
}

@Suppress("unused")
val simSpoofPatch = bytecodePatch(
    name = "Region spoof",
    description = "Adds in-app controls for changing the region TikTok reads.",
    default = true,
) {
    dependsOn(
        regionSpoofHooksPatch,
        settingsPatch,
    )

    compatibleWith(*AppCompatibilities.tiktok4623())

    execute {
        SettingsStatusLoadFingerprint.method.addInstruction(
            0,
            "invoke-static {}, Lapp/morphe/extension/tiktok/settings/SettingsStatus;->enableSimSpoof()V",
        )
    }
}

@Suppress("unused")
val bypassRegionalRestrictionsPatch = bytecodePatch(
    name = "Bypass regional restrictions",
    description = "Uses a selected default region to help bypass regional restrictions.",
    default = false,
) {
    dependsOn(regionSpoofHooksPatch)

    compatibleWith(*AppCompatibilities.tiktok4623())

    val defaultRegion by stringOption(
        key = "defaultRegion",
        title = "Default region",
        description = "Region to use when applying this patch.",
        default = DEFAULT_REGION_PRESET_ID,
        values = regionPresetOptionValues,
        required = true,
    ) { value ->
        value != null && regionPresetOptions.any { it.id == value }
    }

    execute {
        val selectedPreset = regionPresetOption(defaultRegion!!)
        val extensionClass = mutableClassDefBy(classDefBy(EXTENSION_CLASS_DESCRIPTOR))

        extensionClass.staticFields.first { it.name == "DEFAULT_REGION_SPOOF_ENABLED" }.initialValue =
            ImmutableBooleanEncodedValue.forBoolean(true).toMutable()
        extensionClass.staticFields.first { it.name == "DEFAULT_REGION_ISO" }.initialValue =
            MutableStringEncodedValue(ImmutableStringEncodedValue(selectedPreset.iso))
        extensionClass.staticFields.first { it.name == "DEFAULT_REGION_MCC_MNC" }.initialValue =
            MutableStringEncodedValue(ImmutableStringEncodedValue(selectedPreset.mccMnc))
        extensionClass.staticFields.first { it.name == "DEFAULT_REGION_OPERATOR" }.initialValue =
            MutableStringEncodedValue(ImmutableStringEncodedValue(selectedPreset.operatorName))
    }
}
